mono/mini/interp/transform.c

```diff
@@ -7241,7 +7241,7 @@ interp_cprop (TransformData *td)
 			ins = interp_fold_binop (td, sp, ins);
 			sp--;
 		} else if (ins->opcode >= MINT_STFLD_I1 && ins->opcode <= MINT_STFLD_P && (mono_interp_opt & INTERP_OPT_SUPER_INSTRUCTIONS)) {
-			if (sp [-2].ins) {
+			if (sp [-2].ins && sp [-2].val.type != STACK_VALUE_NONE) {
```

Comment:
actually this should probably be `sp [-2].val.type == STACK_VALUE_LOCAL`.

Reply:
Yes, but we also need to still check for sp [-2].ins. Semantically speaking the **ins** field represents the instruction that pushed the value on the stack and the **val** represents what value is on the stack. We can replace with super instruction if the value on the stack is a local (STACK_VALUE_LOCAL) whose push can be removed from the instruction stream (sp [-2].ins != NULL) and we replace it with field access on whatever local the sp [-2].val indicates, rather than accessing the local_index from sp [-2].ins->data [0].

Reply:
note that we still need to handle the ldarg case below

Reply:
> we replace it with field access on whatever local the sp [-2].val indicates, rather than accessing the local_index from sp [-2].ins->data [0].

not sure if I follow. For the local case, shouldn't `g_assert (sp [-2].ins->data [0] == sp [-2].val.local)` be true?

Reply:
That sounds correct but it's harder to follow. I was saying that it would be a good idea to just not access the data of the instruction `sp [-2].ins`, the only reference of this instruction should be when clearing it. This would result in applying the optimization in other theoretical cases like : `ldloc`, `dup`, `stfld` (Now we can apply the superinstruction on dup + stfld, even if dup is not a ldloc, because the stack slot contains the value of a local).

In addition, this bug uncovers the same existing issue on the argument field optimization. If the `ldarg` that is pushed on the stack is later changed by a starg (which should be crazy unrare), then we can no longer apply the optimization. I think we should fix it as part of this same PR. I think fixing it would require to add a new STACK_VALUE_ARG and clear it as we do with locals (when encountering `starg`).

Reply:
Thanks! You are right, the same issue was there with `ldarg`/`starg`. I also added a repro for that scenario.

Fixing this, as you suggested, opened up for more optimization opportunities.  For the repro by Zoltan in the issue (executing Roslyn basically), it increases
* for the locals field optimization: from 110 to 166 fusions
* for the arguments field optimization: from 2134 to 2235 fusions

Reply:
Could you also check before/after for the standard corlib suite (the total amount of super instructions emitted before and after is fine). I'm curious since that is what I used as a reference when adding the optimizations

Reply:
Running `make -C mcs/class/corlib run-test`, before:
```
Interp statistics
Total transform time                : 6520.80 ms
Methods transformed                 : 127963
Total cprop time                    : 553.36 ms
Total super instructions time       : 75.83 ms
STLOC_NP count                      : 107621
MOVLOC count                        : 13007
Copy propagations                   : 24591
Added pop count                     : 27432
Constant folds                      : 72665
Super instructions                  : 271023
Killed instructions                 : 604063
Emitted instructions                : 4597899
Methods inlined                     : 104246
Inline failures                     : 210839
```
After:
```
Interp statistics
Total transform time                : 6579.69 ms
Methods transformed                 : 127969
Total cprop time                    : 532.04 ms
Total super instructions time       : 79.11 ms
STLOC_NP count                      : 107347
MOVLOC count                        : 13016
Copy propagations                   : 24603
Added pop count                     : 27720
Constant folds                      : 72645
Super instructions                  : 273419
Killed instructions                 : 605771
Emitted instructions                : 4602170
Methods inlined                     : 104376
Inline failures                     : 210898
```